theories/lra.elpi

```diff
@@ -0,0 +1,24 @@
+pred field_from_type i:term, o:term.
+field_from_type {{ _ -> lp:Type }} F :- field_from_type Type F.
+field_from_type {{ is_true (@Order.le _ lp:Ty _ _) }} F :-
+  std.assert-ok! (coq.unify-eq {{ Num.RealField.sort lp:F }} {{ Order.POrder.sort lp:Ty }})
+                 "Cannot find a realFieldType".
+
+pred etoQ i:term, i:term, o:term.
+etoQ F {{ lp:E1 + lp:E2 }} {{ (lp:E1Q + lp:E2Q)%Q }} :- etoQ F E1 E1Q, etoQ F E2 E2Q.
+etoQ F {{ lp:E1 * lp:E2 }} {{ (lp:E1Q * lp:E2Q)%Q }} :- etoQ F E1 E1Q, etoQ F E2 E2Q.
+etoQ _F {{ (lp:X)%:R }} XQ :-
+  coq.reduction.vm.norm {{ Qmake (Z.of_nat lp:X) 1 }} {{ Q }} XQ.
+etoQ F {{ lp:X }} {{ @Internals.F2Q lp:F lp:X }}.
+
+pred toQ i:term, i:term, o:term.
+toQ F {{ lp:H -> lp:G }} {{ lp:HQ -> lp:GQ }} :- toQ F H HQ, toQ F G GQ.
+toQ F {{ _ -> lp:G }} {{ lp:GQ }} :- toQ F G GQ.
+toQ F {{ is_true (@Order.le _ lp:O lp:X lp:Y) }} {{ (lp:XQ <= lp:YQ)%Q }} :-
+  coq.unify-eq {{ Num.RealField.sort lp:F }} {{ Order.POrder.sort lp:O }} ok, !,
+  etoQ F X XQ, etoQ F Y YQ.
```

Comment:
>     * `etoQ` ignores the implicit parameter of `GRing.add` and `GRing.mul`. I'd like to check the convertibility between `F` and them.

We could but I don't see how the test could fail if the term is well typed (since we already tested the type of the expression on l.18)

>     * This step can be seen as `preprocessing`. The actual reification happens in `lra`, on the preprocessed formula. I'd like to avoid traversing the input term twice.

I agree we should tweak the code of micromega in Coq so that it directly takes `ff` and `varmap` as input and we avoid doing the reification twice (but we should probably do some bench before, the extra reification may be cheap compared to running the simplex and the checker).

>     * In L.18, I think we should do `coq.unify-eq {{ Num.RealField.porderType lp:F }} O ok` instead.

Just tested, this unfortunately fails.

Reply:
> We could but I don't see how the test could fail if the term is well typed (since we already tested the type of the expression on l.18)

I agree that it is unlikely to happen. But I think it is easy to construct such an example using the converse ring.

Reply:
Sorry, I misread your comment, you are perfectly right. I updated the checks. Thanks!